### PR TITLE
build: add platforms

### LINKED ISSUE #120: Add platform and other properties to service.build
## Build platforms
With the arrival of Arm on developer laptops and in data centers, developers are being pushed into a multi architecture world. Compose services have a platform property of type string which is used to determine what the pulled or built image's platform should be.

This works for local development or deployment as users only need a single architecture at a time. When publishing images to a container registry, users may want to build and push the images for multiple platforms.

**Question:**
Should the build section have a platform property of type list of strings? If so, how does this relate to the service platform?

**Proposal:**
Add a platform string list property to build that is respected only when pushing images, doing an implicit build if needed. This will allow developers to have a quick development cycle (only using one architecture) but allow them to publish their images with multi architecture support.

## Other properties

Investigating further, there may be other build properties that we would like to add. The current set of build properties are:
```json
"context": {"type": "string"},
"dockerfile": {"type": "string"},
"args": {"$ref": "#/definitions/list_or_dict"},
"labels": {"$ref": "#/definitions/list_or_dict"},
"cache_from": {"$ref": "#/definitions/list_of_strings"},
"network": {"type": "string"},
"target": {"type": "string"},
"shm_size": {"type": ["integer", "string"]},
"extra_hosts": {"$ref": "#/definitions/list_or_dict"},
"isolation": {"type": "string"}
```

Docker `buildx` includes the following interesting flags that are currently not exposed in the Compose format:
```console
      --allow strings            Allow extra privileged entitlement, e.g. network.host, security.insecure
      --builder string           Override the configured builder instance
      --cache-from stringArray   External cache sources (eg. user/app:cache, type=local,src=path/to/dir)
      --cache-to stringArray     Cache export destinations (eg. user/app:cache, type=local,dest=path/to/dir)
      --load                     Shorthand for --output=type=docker
      --no-cache                 Do not use cache when building the image
  -o, --output stringArray       Output destination (format: type=local,dest=path)
      --platform stringArray     Set target platform for build
      --secret stringArray       Secret file to expose to the build: id=mysecret,src=/local/secret
      --ssh stringArray          SSH agent socket or keys to expose to the build (format: default|<id>[=<socket>|<key>[,<key>]])
```

**Question:**
Should we add any of these properties?

**Proposal:**
There are some options taht would make sense to add now. Specifically:
- `cache-to`: Logical given that we have `cache-from` already.
- `no-cache`: Logical given the ability to specify a cache.
- `platform`: As described above.
- `secret`: It's a common issue that people do not have an easy way to attach secrets in a way that they are not built into the image.

Some require more consideration as they're either very Docker specific or might be too niche:
- `allow`
- `builder`
- `load`
- `output`
- `ssh`

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**What this PR does / why we need it**:
Add `platforms`  (`[]string`) to `build.md` for supporting multi-platform builds

**Which issue(s) this PR fixes**:

Fixes #120

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @chris-crone @tonistiigi

**Comment by thaJeztah** [comment]
I don't think there was consensus yet on the proposal, because The service under which the build is defined also has a "platform" option https://github.com/compose-spec/compose-spec/blob/master/spec.md#platform

This would be confusing to have, and also means the service has itself can have a different platform than the build defines (?)

**Comment by ndeloof** [comment]
IMHO it makes perfect sense to distinguish platform selected to run a service, and platform(s) we want to support to build the service image, so there's no collision between the existing `service.platform` attribute and the proposed `build.platforms` one.

**Comment by AkihiroSuda** [comment]
> This would be confusing to have, and also means the service has itself can have a different platform than the build defines (?)

Yes, just like `docker run --platform` and `docker build --platform`, and I think that is fine.
An implementation may print a notice message in that case.

**Comment by thaJeztah** [comment]
The `build` here is to a pre-condition for the specified service, not a "generic" build; so what's the expected result of;

```yaml
services:
  foo:
    platform: "linux/amd64"
    build:
      context: .
      platforms: ["linux/amd64", "linux/arm64"]
```

On a daemon only supporting a single architecture in the image store, the above may build `linux/arm64` as last image, overwriting the `linux/amd64`

Similar cases may exist if the `build.platforms` have no intersection with `service.platform`, e.g.;

```yaml
services:
  foo:
    platform: "linux/amd64"
    build:
      context: .
      platforms: ["linux/arm64"]
```

**Comment by AkihiroSuda** [comment]
@thaJeztah Could you take a look again? 🙏

**Comment by glours** [comment]
@ndeloof @EricHripko @justincormack  @ulyssessouza can you take a look please?

**Review by EricHripko — COMMENTED** [review]
Thank you for taking the time to contribute to the spec! I've got a quick question inline 💬

File: build.md [hunk 1]
```diff
@@ -393,6 +393,41 @@ tags:
...
+
+```yml
+build:
+  context: "."
+  platforms:
+    - "linux/amd64"
+    - "linux/arm64"
+```
+
+Compose implementations MUST include the service's platform in the list of the default build target platforms.
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply build.md]: Updated to clarify that

**Review by ndeloof — APPROVED** [review]

File: build.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (glours) [reply build.md]: @EricHripko are Akihiro's modifications good for you? 
That's one of the last missing parts of the build section improvements, I really hope we can move forward soon and close the #233 quickly 😅

**Review by ulyssessouza — APPROVED** [review]

**Comment by glours** [comment]
@AkihiroSuda is your compose-go [PR](https://github.com/compose-spec/compose-go/pull/282) ready to review?

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda is your compose-go [PR](https://github.com/compose-spec/compose-go/pull/282) ready to review?

Yes, please

**Pull request merged**